### DISCUSSION TITLE
How does Wineskin know to use D3DMetal?  (re: D3D DLLs + dylib)

### DISCUSSION THREAD

**Discussion by sofakng** [The-Wineskin-Project/WineskinServer#444] [General]
I'm trying to understand how WINE (Wineskin) knows to use the D3DMetal framework by copying the files to /<wrapper>/Contents/SharedSupport/wine/lib.

If I understand correctly, DXVK replaces the D3D files (d3d10core.dll, d3d11.dll, and dxgi.dll) on the drive_c/windows/system32 directory.  This makes sense because you are changing the DLLs to the DXVK versions on the WINE 'system' drive but D3DMetal seems to have files in a different location?

(I also have a few more questions but this might help me understand better...)

**Comment by Gcenx** [comment]
The “Game porting toolkit” dmg contains files that replace some wine files, these replacement “wine” files work in a similar manner to stock wine files.

`winemac.so` has some modifications to allow it to find D3DMetal.framework so when Apples files are copied into place it can be found and used.

**Reply by sofakng** [reply]
Ahhh OK... I think I understand.

The 'built-in' files from WINE are located in:
- Contents/SharedSupport/wine/lib/wine/x86_64-windows (PE binaries)
- Contents/SharedSupport/wine/lib/wine/x86_64-unix (Unix [dylib] binaries)

Is that correct?

That makes sense and also allows for DXVK to be installed (as 'native' files inside the Contents/SharedSupport/prefix/drive_c/windows/system32 directory) and then toggled.

However, once the GPTK binaries are installed then you can't use WineD3D without removing the GPTK files and restoring the WineD3D files?

**Comment by Gcenx** [comment]
> Is that correct?

Yes.

<br>

You don’t want to mix DXVK with D3DMetal it won’t work correctly.

There’s no uninstalling D3DMetal unless you’ve made a backup of WineD3D files.

**Reply by sofakng** [reply]
I thought you could install DXVK DLLs and then switch them between native and built-in to switch between GPTK and DXVK?

Also, where is winemac.so?  I can't seem to find it.

**Comment by Gcenx** [comment]
> I thought you could install DXVK DLLs and then switch them between native and built-in to switch between GPTK and DXVK?

You _technically_ could until recently but D3DMetal dxgi doesn’t play nicely with DXVK dlls and DXVK dxgi isn’t great on macOS where we prefer wines dxgi.

> Also, where is winemac.so? I can't seem to find it.

For wine-7.7 it’s a probably named `winemac.drv.so`, in wine-8.x it’s `winemac.so`

**Reply by sofakng** [reply]
> You _technically_ could until recently but D3DMetal dxgi doesn’t play nicely with DXVK dlls and DXVK dxgi isn’t great on macOS where we prefer wines dxgi.

That makes sense not to combine D3DMetal dxgi with DXVK dxgi.  However, I'm confused about your comment 'we prefer wines dxgi'.  Are you saying it's not a good idea to install wine tricks dxvk_master_dxgi? (in a DXVK setup)

> For wine-7.7 it’s a probably named `winemac.drv.so`, in wine-8.x it’s `winemac.so`
I see it now.  Thanks!  I'm guessing this is also needed for Crossover to find the GPTK files?

Sorry for so many questions and I really appreciate your help!  It's great to understand the inner-workings to help troubleshoot and experiment, etc.

**Reply by Gcenx** [reply]
> That makes sense not to combine D3DMetal dxgi with DXVK dxgi. However, I'm confused about your comment 'we prefer wines dxgi'. Are you saying it's not a good idea to install wine tricks dxvk_master_dxgi? (in a DXVK setup)

Yes it’s not a good idea to install that verb, it only exists to allow DXVK to work when D3DMetal was integrated though it’s still not a good idea. 

>  I see it now.  Thanks!  I'm guessing this is also needed for Crossover to find the GPTK files?

For CXPatcher we’re forcing DXVK dxgi but that causes other issues.

**Reply by sofakng** [reply]
Oh that’s interesting… You can’t use the GPTK DXGI with Crossover?

If so, does that mean Wineskin + CX engine + GPTK is better than Crossover + GPTK?

**Reply by Gcenx** [reply]
> Oh that’s interesting… You can’t use the GPTK DXGI with Crossover?

Maybe I didn’t explain well?, D3DMetal requires it’s own DXGI dll, DXVK can use wines DXGI or DXVK dxgi dll.

CrossOver when DXVK is toggled copies on the DXVK dlls, when disabled the dlls are reverted. For CXPatcher I have it force use DXVK dxgi when available. When a native dxgi isn’t found it’ll use D3DMetal dxgi or wines.

> If so, does that mean Wineskin + CX engine + GPTK is better than Crossover + GPTK?

Wineskin has many more additional patches/fixes/hacks than CXPatcher as it would become way too complex to simply patch.

**Reply by sofakng** [reply]
OK - Sorry.  The confusion was in the statement of 'For CXPatcher we’re forcing DXVK dxgi but that causes other issues.'

That makes me think you are using DXVK DXGI with GPTK, but I think you mean that you are hijacking the Crossover DXVK toggle to implement GPTK, right?

**Reply by Gcenx** [reply]
On when DXVK toggled CrossOver will have copied all of DXVK files into place this include it’s own dxgi dll.

When DXVK is toggled off CrossOver removes DXVK dlls and replaces them with wines builtin dlls.

For CXPatcher this mean;
- DXVK on mean using all its dlls
- DXVK off mean using D3DMetal dlls
